### PR TITLE
Adding ECDS config dump support.

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
For issue: https://github.com/envoyproxy/envoy/issues/20086

Signed-off-by: Yanjun Xiang <yanjunxiang@google.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/common/filter/config_discovery_impl.h
Comment (adisuissa) [inline source/common/filter/config_discovery_impl.h:537]: I wonder how much memory this may end up using, especially if not used.
@jmarantz from your experience on the admin component, is memory consumption of saving raw protobufs something that we should take into consideration?
If so, will explicitly enabling/disabling the ECDS filter tracking the right way to go?

**Review by adisuissa — COMMENTED** [review]
Thanks!
Overall the API LGTM.
2 high level comments/questions:
1. Memory consumption could be non-negligible here. Is there anything we can do to only impact those that want to use this feature?
2. Should there be a way to filter for a specific ECDS config (e.g., using https://www.envoyproxy.io/docs/envoy/latest/operations/admin#get--config_dump?name_regex=)?

File: api/envoy/admin/v3/config_dump_shared.proto
Comment (yanjunxiang-google) [reply api/envoy/admin/v3/config_dump_shared.proto]: Done for the comments change!

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [reply source/common/filter/config_discovery_impl.h:537]: This is a substantial concern for Wasm filters (which can be O(10MB) each). Although it's probably marginal on top of v8 overhead, but it does add O(10MB) at rutnime.

File: source/common/filter/config_discovery_impl.cc
Comment (kyessenov) [inline source/common/filter/config_discovery_impl.cc:134]: Could we avoid duplicating `filter_config`? I think `message` is the same content?

Comment (kyessenov) [reply source/common/filter/config_discovery_impl.cc:134]: Just to add more context: I think there is one subscription on the main thread, so we can probably reuse that one for the config dump. That helps reducing overhead with more copies of the config.

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [inline source/common/filter/config_discovery_impl.h:415]: If we directly pack lastConfig() into config_dump any filed, then config dump doesn't have the filter name info. It makes it hard to know what's the filter name. Thus we first put the lastConfig() + name into this typed extension proto.

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [inline source/common/filter/config_discovery_impl.h:401]: the config dump callbacks are stored in a keyed hashmap, thus the keys has to be different to avoid one override the other. Thus we need getConfigDumpType() to get different strings for each filter type, like HTTP filter, TCP listener filter,...

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [inline test/integration/extension_discovery_integration_test.cc:881]: If both HTTP ECDS filter and listener ECDS filters are configured, then there will be two EcdsConfigDump entries in the /config_dump.  Since the data are stored in two different objects, there is no way to dump them into one EcdsConfigDump proto.
1) For listener ECDS filter: https://github.com/envoyproxy/envoy/blob/b0d79665240cdfe3d87280283d28ac30334542e9/source/server/listener_manager_impl.h#L111

2) For HTTP ECDS filter:
https://github.com/envoyproxy/envoy/blob/b0d79665240cdfe3d87280283d28ac30334542e9/source/extensions/filters/network/http_connection_manager/config.cc#L224

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h]: Understood

File: source/common/filter/config_discovery_impl.cc
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.cc]: Yes, that's a great idea! The latest code is directly to dump the data in the already existing subscriptions_ hash map: https://github.com/envoyproxy/envoy/blob/b0d79665240cdfe3d87280283d28ac30334542e9/source/common/filter/config_discovery_impl.h#L392

With this, there is no new memory consumption added by this PR.

**Comment by yanjunxiang-google** [comment]
@envoyproxy/api-shepherds PTAL

File: test/integration/listener_extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/listener_extension_discovery_integration_test.cc:581]: Same as above.

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [inline source/common/filter/config_discovery_impl.h:608]: Suggestion: `statPrefix` seems like it accomplishes the same goal of partitioning configs in the stats hierarchy. Maybe reuse?

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [inline source/common/filter/config_discovery_impl.h:417]: Is lastConfig always non-null?

File: source/common/filter/config_discovery_impl.h
Comment (kyessenov) [inline source/common/filter/config_discovery_impl.h:410]: Maybe check that weak_ptr lock is non-null for defensive programming?

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:886]: The thing is that ECDS config dump is only setup after ECDS filter is actually configured, which is here.  And we don't want to duplicate the ECDS filter configurations into other generic files as mentioned.

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:858]: This test, "BasicSuccessConfigDump", has only one Listener ECDS filter configured, which is to test the basic case.

"TwoSubscriptionsDifferentNameWithConfigDump" has two listener filter configured. This test can give us a little bit more assurance that the function dumpEcdsFilterConfigs() can handle multiple ECDS filtersconfigured.

Sorry, maybe I don't fully understand the comments.  Does this answer your question?

File: test/integration/extension_discovery_integration_test.cc
Comment (adisuissa) [reply test/integration/extension_discovery_integration_test.cc:886]: Hmmm... 
Ok, so why is the ECDS test checking for "ClustersConfigDump"?
Will it make sense to only check that the 2 "Ecds" types are part of the config dump?

File: test/integration/extension_discovery_integration_test.cc
Comment (adisuissa) [reply test/integration/extension_discovery_integration_test.cc:858]: Maybe I'm misunderstanding something here.
I see the use of `sendXdsResponse2` which I assumed was for a second stream, no?

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:309]: Ok, let me change the name into sendListenerFilterEcdsResponse(). These two functions have significantly different parameters. Let's just keep them separated.

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:886]: Yes, we can do something like below to filter out others to only keep ECDS config dump.
EXPECT_EQ("200", request("admin", "GET", "/config_dump?resource=ecds_filters", response));

However, I feel the full dump can give us the idea where and how these ECDS config dumps are stored and dumped.  BTW, these configure dumps are dumped in the order of the keys associated with it.  For example, we intentionally set the keys to something like "ecds_filter_http", ecds_filter_tcp_listener_filter" to keep them together. If we put them as "http_ecds" and "listener_ecds", they will be spreaded around with other type of config dumps, and maybe hard to find.

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h:608]: A couple of issue if we reuse statPrefix as the config dump key:

1) Is that naming confusing?
2) Upstream UpstreamHttpFilterConfigProviderImpl has identical stats prefix with HttpFilterConfigProviderImpl. If we go that route, we need to modify them to avoid one config dump override the other.
3)  stats prefix are named as "http", or "tcp", if we use them as config dump keys, it will cause ECDS config dump spreaded around and mixed with other types since config dumps are dumped in the alphabetic order of the keys.

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:858]: Oh, sorry, I miss read the comments.

Anyway, the tests are written for below purpose:

1) The BasicSuccenssConfigDump in listener_extensions_discovery_integration_test.cc , covers the basic  ECDS filter config dump case, only one ECDS filter is configured. 

2)  The TwoSubscriptionsDifferentNameWithConfigDump in listener_extensions_discovery_integration_test.cc , covers the config dump case for two  ECDS filters with same type .

3) The BasicSuccenssConfigDump in extensions_discovery_integration_test.cc , covers the config dump test for two ECDS filters with different type case: one is HTTP, the other is Listener filter.

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h]: Good catched. Added a check.

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h:410]: Good idea. Done!

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h:610]: Yeah, filed an issue to track 2: https://github.com/envoyproxy/envoy/issues/24010

**Review by kyessenov — COMMENTED** [review]
Code LGTM, please wait for other to approve.

**Comment by yanjunxiang-google** [comment]
> Code LGTM, please wait for @adisuissa to approve.

Thanks @kyessenov !

I pushed one more commits to add the ECDS config_dump documentation. Also added another test to verify ECDS config_dump works well with filter name based regex lookup.  The change is small.   PTAL!

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this! Left a minor comment on the test and one nit, but otherwise LGTM.
/assign-from @envoyproxy/senior-maintainers

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc]: Thanks for the comments.

Let me change the test name of (3) to reflect it's testing two ECDS types (HTTP filter and Listener filter).  The reason we need this test is that we what to see if both types of ECDS filters are configured, the config dump actually looks like below, i.e, two ECDS filters are dumped back-to-back after "cluster" and consecutively:

  // * ``bootstrap``: :ref:`BootstrapConfigDump <envoy_v3_api_msg_admin.v3.BootstrapConfigDump>`
  // * ``clusters``: :ref:`ClustersConfigDump <envoy_v3_api_msg_admin.v3.ClustersConfigDump>`
  // * ``ecds_filter_http``: :ref:`EcdsConfigDump <envoy_v3_api_msg_admin.v3.EcdsConfigDump>`
  // * ``ecds_filter_tcp_listener``: :ref:`EcdsConfigDump <envoy_v3_api_msg_admin.v3.EcdsConfigDump>`
  // * ``endpoints``:  :ref:`EndpointsConfigDump <envoy_v3_api_msg_admin.v3.EndpointsConfigDump>`
  // * ``listeners``: :ref:`ListenersConfigDump <envoy_v3_api_msg_admin.v3.ListenersConfigDump>`
  // * ``scoped_routes``: :ref:`ScopedRoutesConfigDump <envoy_v3_api_msg_admin.v3.ScopedRoutesConfigDump>`

If we split these in two separate tests, like one for HTTP, one for ECDS, we will only see one EcdsConfigDump here in each test.  Please check the added documentation below:

 // Currently ECDS is supported in HTTP and listener filters. Note, ECDS configuration for
  // either HTTP or listener filter will only be dumped if it is actually configured.
  
For 2), it is test with only one ECDS filter typed configured(here is Listener, but it's same for HTTP), but there are multiple such filters, the EcdsConfigDump are working fine. So what we saw one EcdsConfigDump entry like below in Gete /config_dump. But inside EcdsConfigDump, there are two filters got dumped.

  // * ``bootstrap``: :ref:`BootstrapConfigDump <envoy_v3_api_msg_admin.v3.BootstrapConfigDump>`
  // * ``clusters``: :ref:`ClustersConfigDump <envoy_v3_api_msg_admin.v3.ClustersConfigDump>`
  // * ``ecds_filter_tcp_listener``: :ref:`EcdsConfigDump <envoy_v3_api_msg_admin.v3.EcdsConfigDump>`
  // * ``endpoints``:  :ref:`EndpointsConfigDump <envoy_v3_api_msg_admin.v3.EndpointsConfigDump>`
  // * ``listeners``: :ref:`ListenersConfigDump <envoy_v3_api_msg_admin.v3.ListenersConfigDump>`
  // * ``scoped_routes``: :ref:`ScopedRoutesConfigDump <envoy_v3_api_msg_admin.v3.ScopedRoutesConfigDump>`

This is to verify the for loop in the added function dumpEcdsFilterConfigs() work as expected.

Also, the reason we add the two ECDS types test in extension_discovery_integration_test.cc, not in listener _extension_discovery_integration_test.cc is that it's much easier to add a listener ECDS filter configuration into a test already has HCM and HTTP ECDS filter, which is the case of extension_discovery_integration_test.cc . And it needs much bigger changes to add HCM and HTTP ECDS filter into a test has Listener ECDS filter (listener _extension_discovery_integration_test.cc).

File: source/common/filter/config_discovery_impl.h
Comment (yanjunxiang-google) [reply source/common/filter/config_discovery_impl.h:396]: done

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [inline test/integration/extension_discovery_integration_test.cc:860]: With https://github.com/envoyproxy/envoy/pull/23903, the ADMIN interface is disabled in compile_time_options CI build. Have to disable these tests in this case since config_dump requires admin interface.

**Comment by yanjunxiang-google** [comment]
@envoyproxy/api-shepherds  could you please approve this PR?

File: api/envoy/admin/v3/config_dump.proto
Comment (adisuissa) [inline api/envoy/admin/v3/config_dump.proto:36]: Should `ecds_filter_upstream_http` and `ecds_filter_udp_listener` also be added here, or is it because they are not plumbed at the moment?

Comment (yanjunxiang-google) [reply api/envoy/admin/v3/config_dump.proto:36]: That's a good question.  ecds_filter_udp_listener is not plumbed yet so we should leave it out.

ecds_filter_upstream_http leveraged most of the ecds_filter_http code, so theoretically it should work automatically.  However, I don't see any Upstream HTTP filter related test code thus guess it is WIP.  Let's leave it out until the feature is ready and related tests are in place.

**Comment by yanjunxiang-google** [comment]
@envoyproxy/api-shepherds Any comments on the API? Otherwise, could you please approve?

File: test/integration/extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/extension_discovery_integration_test.cc:262]: Previously, for LEDS, I remember I had a race between the 2 incoming ECDS streams. When registering for 2 resources, sometimes resource-A was received on stream-1, and sometimes resource-B was received on stream-1. Did you encounter something similar on the test that uses 2 streams?

One way to check (though not guaranteed) is to run the test 1000 times and see if it is flaky.

File: test/integration/extension_discovery_integration_test.cc
Comment (adisuissa) [inline test/integration/extension_discovery_integration_test.cc:876]: Are they both sending xDS responses, but one on stream1 and the other on stream2?
If so, I suggest rewriting the method names/generalizing them, so it will be more clear.

**Review by adisuissa — COMMENTED** [review]
Thanks!
/lgtm api

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc:262]: In this case, the the two ECDS subscription are on two completely different connections. Each subscription has it's own cluster and upstream. I guess that's the reason it doesn't race.  BTW the logic is also in https://github.com/envoyproxy/envoy/blob/dede92e6e5888cdf27320bf3b6bc9987657d6e22/test/integration/listener_extension_discovery_integration_test.cc#L130, which had been in code base for a while.

I run the 1000 tests anyway, and doesn't flaky.

File: test/integration/extension_discovery_integration_test.cc
Comment (yanjunxiang-google) [reply test/integration/extension_discovery_integration_test.cc]: done

**Review by kyessenov — APPROVED** [review]

**Comment by alyssawilk** [comment]
I believe this was the other half of breaking coverage for main
Again I'd be inclined to roll back and roll forward when it has sufficient testing of new code
https://github.com/envoyproxy/envoy/pull/24320

**Comment by yanjunxiang-google** [comment]
> I believe this was the other half of breaking coverage for main Again I'd be inclined to roll back and roll forward when it has sufficient testing of new code #24320

When working on #23902 , we found the code : UpstreamHttpFilterConfigProviderManagerImpl in config_discovery_impl.h is not tested, 
https://github.com/envoyproxy/envoy/blob/5b9183f80d6c7ab38877b8f8f48b152509f49d3e/source/common/filter/config_discovery_impl.h#L578

which is causing the test coverage in that file is low.

**Comment by alyssawilk** [comment]
Coverage passed on main before your PR, and your PR made additions to a file.  I have never seen this then fail without the reason being your new code lowered overall coverage.  I'm sorry about this mix up and happy to point you at the coverage results when you try to reinstate it and help you sort out how to do it.  The greater problem was this PR was unintentionally merged without the correct approvals so we have to fix that up as well.

**Comment by alyssawilk** [comment]
yeah looks like you were penalized for 2 comments (stupid sorry) but also didn't test an early return: https://storage.googleapis.com/envoy-postsubmit/main/coverage/source/common/filter/config_discovery_impl.h.gcov.html

**Comment by yanjunxiang-google** [comment]
Hi, @alyssawilk , I quickly created this PR to address the test flaky issue in config_discovery_impl.h: https://github.com/envoyproxy/envoy/pull/24362

**Pull request merged**
